lib/gpu/render_pass.cc

```diff
@@ -142,8 +142,15 @@ RenderPass::GetOrCreatePipeline() {
 
   std::shared_ptr<impeller::Pipeline<impeller::PipelineDescriptor>> pipeline;
 
-  if (context.GetBackendType() == impeller::Context::BackendType::kOpenGLES) {
-    // Get the pipeline on the raster thread.
+  if (context.GetBackendType() == impeller::Context::BackendType::kOpenGLES &&
+      !context.GetPipelineLibrary()->HasPipeline(pipeline_desc)) {
+    // For GLES, new pipeline creation must be done on the reactor (raster)
+    // thread. We're about the draw, so we need to synchronize with a raster
+    // task in order to get the new pipeline. Depending on how busy the raster
+    // thread is, this could hang the UI thread long enough to miss a frame.
+
+    // Note that this branch is only called if a new pipeline actually needs to
+    // be built.
     auto dart_state = flutter::UIDartState::Current();
```

Comment:
So I think this is a problem in the HAL design. We expect to be able to have a pipeline synchronously for encoding but 1) we can only get the pipeline on a particular thread for GLES and 2) we don't _actually_ need the pipeline until we render - which always happens on the react thread.

I think it would be something to punt to a new PR, but I would consider change the HAL design such that SetPipeline accepted the pipleine future. The Vulkan/Metal implementation would call waitandget while the GLES implementation would defer that until encoding.

Reply:
Yeah I completely agree with this. Started filing an issue for it but you beat me to it. :)